### ISSUE TITLE
Move shape-flattening from dataset preprocessing into a model layer

### ISSUE THREAD

**Issue by LeoBuron** [es-ude/OnDeviceTraining#76]
## Problem

In `experiments/MnistExperiment.c`, the function `flattenItemDims()` (lines 91–139) mutates each sample tensor's `shape_t` in place to collapse image data from `[1, 28, 28]` to `[1, 784]` so the downstream `Linear` layer can consume it. It is invoked between `initDataSets()` and `dataLoaderInit()` (call site: line 212).

This has four concrete issues:

### 1. Latent bug — only works for square inputs

`MnistExperiment.c:104` (and `:124`) computes the flattened dimension as:

```c
size_t newDimsData[] = {shape->dimensions[0],
                        shape->dimensions[2] * shape->dimensions[2]};
```

This squares `dim[2]` instead of multiplying `dim[1] * dim[2]`. The result happens to be correct for MNIST only because `H == W == 28`. Any non-square input (e.g. `[1, 3, 32]` or rectangular images) would silently produce a wrong flattened size with no assertion failure — the DataLoader and Linear layer would just operate on garbage shapes.

### 2. Wrong abstraction layer

Shape transformations are a model concern, not a dataset concern. PyTorch, Keras, and the IR in `elastic-ai.creator` all place `Flatten`/`Reshape` as the first layer of the model, not as a dataset preprocessing step. The current placement:

- Couples the dataset to the downstream model topology (any dataset feeding a 2D model must flatten).
- Prevents a single dataset from being reused with different model architectures (e.g. a Conv1d model that expects `[C, L]`).
- Spreads shape handling across layers — `MnistExperiment.c` owns some of it, `src/layer/` owns the rest.

### 3. Blocks ir2c code generation

The annotated IR produced by `elastic-ai.creator` represents `Flatten` as an operator in the model graph. `ir2c` is expected to emit a corresponding C layer. There is no IR construct for "dataset-side shape mutation," so the current approach has no path through the compilation pipeline from PyTorch → annotated IR → C code.

### 4. Current state — no Flatten/Reshape infrastructure exists

A grep across `src/` shows zero hits for `flatten`, `reshape`, or `view`. The only shape-manipulation primitive is `transposeTensor()` (`src/tensor/Tensor.c:236-243`), which is zero-copy via `orderOfDimensions` permutation. There is no existing building block for producing a 2D view of an N-D tensor.

## Scope

Solve the problems above by:

- Adding a proper `FLATTEN` layer to the C framework.
- Removing `flattenItemDims()` from `MnistExperiment.c`.
- Wiring the Flatten layer as the first layer of the MNIST model.
- Documenting the convention in `CLAUDE.md`: *datasets deliver samples in their natural geometric shape; any reshape/flatten is the first layer of the model*.

The dimension-bug (`dim[2]*dim[2]`) is resolved implicitly by deleting `flattenItemDims()`.

## Proposed approach

Staged in two PRs to separate architectural change from performance optimization.

### Stage 1 — non-in-place Flatten (this issue, core PR)

- Add `FLATTEN` to `layerType_t` enum in `src/layer/include/Layer.h`.
- Add `flattenConfig_t` (fields: `quantization_t *forwardQ`, `quantization_t *backwardQ`) to the `layerConfig_t` union.
- Implement `src/layer/Flatten.c` with `flattenForward`, `flattenBackward`, `flattenCalcOutputShape` following the ReLU template (`src/layer/Relu.c` + `include/Relu.h`).
- Forward: `memcpy(output->data, input->data, size)` and set output shape to `[batch, product(input.dims[1:])]`.
- Backward: `memcpy(propLoss->data, loss->data, size)` and restore original input shape.
- Register in `layerFunctions[]` vtable (`src/layer/Layer.c:9-14`).
- Add userApi factory in `src/userApi/layer/FlattenApi.c` following the `ReluApi.c` pattern.
- Refactor `experiments/MnistExperiment.c`: remove `flattenItemDims()`, insert Flatten as `model[0]`.
- Add dataset-shape convention to `CLAUDE.md`.
- Unit tests in `test/unit/layer/UnitTestFlatten.c` — include a non-square input case (e.g. `[1, 3, 4, 5]` → `[1, 60]`) as a regression test for the old `dim[2]*dim[2]` bug class.

**Why non-in-place first:** consistent with every existing layer (memcpy pattern). Low review burden. Delivers the architectural fix, the bug fix, the ir2c-compatibility, and the convention in one reviewable chunk.

### Stage 2 — in-place Flatten variant (separate follow-up issue)

A zero-copy variant where `output->data` aliases `input->data` with a reshaped `shape_t`. This is a pure performance optimization (saves one `memcpy` per batch per forward/backward pass) but collides with the output-preallocation in `initLayerOutputs()` (`src/userApi/training_loop/calculate_grads/CalculateGradsSequential.c:75-124`), which expects every layer to own a distinct output buffer. Resolving this cleanly (skip-allocation path vs. allocate-then-overwrite vs. a new `isAlias` flag on layers) is a design discussion of its own and deserves its own issue.

**Why split:** architecture vs. performance have different review criteria. Mixing them in one PR drags review time. The `initLayerOutputs` change is the riskier part and should not gate the architectural fix.

## Out of scope

- General `Reshape`/`View` layer — no use case identified beyond flatten-to-2D. DenseNet-style concat is a different operation handled separately.
- Tensor-level reshape helpers (`tensorReshape`/`tensorView`) — the layer encapsulates shape handling.
- Changes to Linear layer or DataLoader interfaces.

## Acceptance criteria

- [ ] `flattenItemDims()` removed from `experiments/MnistExperiment.c`.
- [ ] `FLATTEN` layer type registered in `layerType_t` enum and `layerFunctions[]` vtable.
- [ ] Flatten layer wired as `model[0]` in `MnistExperiment`; training converges to accuracy comparable to the previous run.
- [ ] Unit tests cover a non-square input shape (to guard against the regression class of the `dim[2]*dim[2]` bug).
- [ ] `CLAUDE.md` contains the dataset-vs-model shape convention.

## Related

- Blocks: Layer one-liner issue (follow-up) — a `flattenLayerInit*` factory is prerequisite for one-liner work.
- Follow-up: a separate issue will cover the in-place (zero-copy) Flatten variant and the `initLayerOutputs` refactor it requires.

**Comment by LeoBuron** [comment]
Fix lives on `develop` (commits 7afbd2f, d83dadf, 2387632, 0d7d485, 125132e, 5cda316, 74378ca, 8565b19, 1e7dd06, 8e8f077, plus docs commits). Tracking PR #77 → `main` for the merge — will close this issue once #77 is merged.

**Comment by LeoBuron** [comment]
Fixed on `develop` (FLATTEN layer Stage 1 — commits 7afbd2f…8e8f077 plus docs commits). PR #77 → `main` remains open for propagation. Closing manually per the develop-only policy; will not be reopened when #77 merges.
